Support pipelining (batching) of mutating ops to same bucket

vekterli opened the pull request:
@baldersheim please review

Bucket operations require either exclusive (single writer) or shared (multiple readers) access. Prior to this commit, this means that many enqueued feed operations to the same bucket introduce pipeline stalls due to each operation having to wait for all prior operations to the bucket to complete entirely (including fsync of WAL append). This is a likely scenario when
feeding a document set that was previously acquired through visiting, as such documents will inherently be output in bucket-order.

With this commit, a configurable number of feed operations (put, remove and update) bound for the exact same bucket may be sent asynchronously to the persistence provider in the context of the _same_ write lock. This mirrors how merge operations work for puts and removes.

Batching is fairly conservative, and will _not_ batch across further messages when any of the following holds:

 * A non-feed operation is encountered
 * More than one mutating operation is encountered for the same document ID
 * No more persistence throttler tokens can be acquired
 * Max batch size has been reached

Updating the bucket DB, assigning bucket info and sending replies is deferred until _all_ batched operations complete.

Max batch size is (re-)configurable live and defaults to a batch size of 1, which shall have the exact same semantics as the legacy behavior. A later follow-up will add a feature flag to control this.

Additionally, clock sampling for persistence threads have been abstracted away to allow for mocking in tests (no need for sleep!).

vekterli committed:
Support pipelining (batching) of mutating ops to same bucket

Bucket operations require either exclusive (single writer) or
shared (multiple readers) access. Prior to this commit, this
means that many enqueued feed operations to the same bucket
introduce pipeline stalls due to each operation having to wait
for all prior operations to the bucket to complete entirely
(including fsync of WAL append). This is a likely scenario when
feeding a document set that was previously acquired through
visiting, as such documents will inherently be output in
bucket-order.

With this commit, a configurable number of feed operations
(put, remove and update) bound for the exact same bucket may
be sent asynchronously to the persistence provider in the
context of the _same_ write lock. This mirrors how merge
operations work for puts and removes.

Batching is fairly conservative, and will _not_ batch across
further messages when any of the following holds:

 * A non-feed operation is encountered
 * More than one mutating operation is encountered for the
   same document ID
 * No more persistence throttler tokens can be acquired
 * Max batch size has been reached

Updating the bucket DB, assigning bucket info and sending
replies is deferred until _all_ batched operations complete.

Max batch size is (re-)configurable live and defaults to a
batch size of 1, which shall have the exact same semantics as
the legacy behavior.

Additionally, clock sampling for persistence threads have been
abstracted away to allow for mocking in tests (no need for sleep!).

vekterli commented:
Seems to be some nondeterminism in a test, will try to reproduce and remove.

vekterli committed:
Ensure visibility of max batch size reconfiguration in persistence thread

baldersheim commented on storage/src/vespa/storage/persistence/filestorage/filestorhandlerimpl.cpp:
```diff
@@ -992,6 +1008,92 @@ FileStorHandlerImpl::Stripe::getNextMessage(vespalib::steady_time deadline)
+    case api::MessageType::REMOVE_ID:
+        return dynamic_cast<const api::RemoveCommand&>(msg).getDocumentId().getGlobalId();
+    case api::MessageType::UPDATE_ID:
+        return dynamic_cast<const api::UpdateCommand&>(msg).getDocumentId().getGlobalId();
```
dynamic_cast is a rather expensive operation. Consider replacing with static_cast or a getAsXXMsg cast safe method.

vekterli replied:
Replaced with `static_cast`. Downcast-safe type invariant shall be maintained by the message's own type ID tracking. If it's not, we'd have other, bigger problems.

baldersheim commented:
PR build has spotted something I did not  .....

vekterli committed:
Add Abseil failure signal handler to test runner to get stack dumps for crashes

vekterli committed:
Rewrite test to manually start single persistence thread

Avoids the need for barriers to avoid stepping on the thread's toes

vekterli committed:
Use `static_cast` instead of `dynamic_cast`

Downcast-safe type invariant shall be maintained by the message's
own type ID tracking. If it's not, we have bigger problems.

vekterli committed:
Ensure all async reply processing executor tasks have completed

vekterli committed:
Low-level message fetch routine must not implicitly unlock mutex

Implicitly unlocking messes up higher level assumptions about
when locks are held and thus cannot be safely done. Lock will
be unlocked immediately after anyway, so this does not seem
like a useful optimization.

toregge commented:
Running all unit tests in parallel causes the following failure on x86_64 :

```
gdb) where
#0  raise (sig=11) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  <signal handler called>
#2  0x00007f0895e556aa in storage::MessageTracker::sendReply (this=0x7f0874001e80)
    at /home/tegge/git/vespa/storage/src/vespa/storage/persistence/persistenceutil.cpp:155
#3  0x00007f0895e3ad31 in operator() (response=..., __closure=0x7f0874000da8)
    at /opt/rh/gcc-toolset-13/root/usr/include/c++/13/bits/unique_ptr.h:199
#4  storage::(anonymous namespace)::LambdaResultTask<storage::AsyncHandler::handlePut(storage::api::PutCommand&, storage::MessageTracker::UP) const::<lambda(storage::spi::Result::UP)> >::run(void) (this=0x7f0874000d90)
    at /home/tegge/git/vespa/storage/src/vespa/storage/persistence/asynchandler.cpp:66
#5  0x00007f088c84a4ba in vespalib::ThreadStackExecutorBase::run (this=0x20ee000)
    at /home/tegge/git/vespa/vespalib/src/vespa/vespalib/util/threadstackexecutorbase.cpp:125
#6  0x00007f0896711dd9 in storage::(anonymous namespace)::test_executor (worker=...)
    at /home/tegge/git/vespa/storage/src/tests/common/teststorageapp.cpp:96
#7  0x00007f0888220b23 in std::execute_native_thread_routine (__p=0x20ee5c0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#8  0x00007f08884fb1ca in start_thread (arg=<optimized out>) at pthread_create.c:479
#9  0x00007f0887838e73 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) 
```

vekterli committed:
Ensure SetUp/TearDown symmetry with test superclass

Can't initialize members in constructor that depend on objects
that are subsequently reset by the superclass' `SetUp()` method.

vekterli committed:
Ensure parameter evaluation order does not have side effects

vekterli commented:
A few sins have been absolved... Some new, some old.
